### PR TITLE
[WIP] Developer docs

### LINKED ISSUE #114: [Docs] Design Decisions
# ISA/ARC Specification

- All non term columns MUST be unique. (Input, Output, ProtocolREF, ProtocolType) 
- All term columns MAY be used any number of times.
- All table names inside one assay must be unique.

# Function

- ArcTable functions MUST always result in a valid table!
- All table body fields MUST be added explicitly to `ArcTable.Values`.
  - Therefore, adding a column with 2 rows to a table of 5 rows MUST also fill the missing 3 rows with "empty cells".
  - Therefore, adding a column with 7 rows to a table of 5 rows MUST also fill the missing 2 rows for all other columns in the table with "empty cells".
- Any function getting a value from ArcTable (row, column, cell) outside of table range MUST raise an exception or return `None`.
- Any function adding a value to ArcTable MUST only allow indices in table range + 1 (which MUST result in add new and append logic).
- Any "SingleColumn", such as Input, ProtocolREF... MUST have all `CompositeCell.Freetext` cells.
- Any "TermColumn", such as Factor, Parameter, plus "FeaturedColumn"s (ProtocolType) MUST contain `CompositeCell.Term` or `CompositeCell.Unitized`. These columns MAY be uniform or mixed, but MUST NOT contain any `CompositeCell.Freetext`.
- Any member function starting with the set keyword (`SetCellAt` , `SetColumn`,..) MUST NOT add new elements but only replace existing.

# Syntax

- All member functions (table.AddColumn,..) MUST return unit and work on mutable elements.
- All member functions (table.AddColumn,..) MUST have a equivalent static member function of the same name with lower-case first letter (ArcTable.addColumn).
- All static members on ArcTable (ArcTable.addColumn) MUST implement its equivalent member function on a copy of the given table. Therefore, not changing the existing table, but returning a new one!
```fsharp
// example
static member addColumn (header: CompositeHeader,cells: CompositeCell [],?Index: int ,?ForceReplace : bool) : (ArcTable -> ArcTable) =
   fun (table: ArcTable) ->
        let newTable = table.Copy()
        newTable.AddColumn(header, cells, ?index = Index)
        newTable
```
- All CRUD functions between assay and study are implemented on Investigation level. Propagated functions on assay/study must use investigation functions to avoid code duplication.

### LINKED ISSUE #223: [Docs] Setup Minimum Documentation
- [x] Code examples and explanation for loading libraries
- [ ] Code examples and explanation for IO and contracts
- [ ] Code examples for adding and mutating assays, tables and columns

### LINKED ISSUE #231: [BUG] SetISAFromContracts `enableLogging` flag does nothing
image

unused param `enableLogging`

### LINKED ISSUE #232: [Feature Request] Reduce code duplication by inheriting ArcTables() in Assay and Study
**Is your feature request related to a problem? Please describe.**

Currently we initiate a ArcTables type for every function related to `ResizeArray<ArcTable>` calling the ArcTables function on the fresh type. This lead to quite some code duplication and is difficult to maintain, as it is easy to miss reimplementing certain functions

**Describe the solution you'd like**

Inherit the ArcTables type as seen in the following FABLE REPL.

https://fable.io/repl/#?code=LYewJgrgNgpgBAYSgQwM6oLAChsgA4wB2cAYsgEawB0CIATjNtgNoA8AggC6fIDGAFgFkYwcjDqoAfAF1snAJ4E4ASQAqimAAoAlHAC8cbHGNxgIsXTgA3ZFEQgIhTvrgAGOAHcAlp35wA5jCcADRwqEFMWGxcPALCouJSslgKSoLy6gQ6+kYmXoT84j4qmVrakSam5uJwvl6oVMqEvAxoZTlYlZV1DbSOzqwAtLX89TQOTnAA1HAAjBUmZgmWPVQAIjAtMG3ZerldI2N9k0OHvRPOw-M4nYvVK6MNAMr8IB67cHh0+ZwAZsQAIgQEDoDEmvAucHqAC44ABSLwAs7jfqRWDOAAeLnSpR0kQxjWarXCeJu6LgvBcBOOzgA9LSzmFXtAwHAxHBqZD8nAAFaYG4El5vHRAA&html=Q&css=Q

With this solution we would only need to implement any function once in ArcTables, then inherit ArcTables with the given tables in ArcAssay and ArcStudy. Therefore, having all functions on both elements, without further work.

### LINKED ISSUE #233: [BUG] Remove reference to removed developer branch in gh-actions
image

### LINKED ISSUE #234: [BUG] Wrong error message in template read
Found in Swate-templates testing. 

Should be: no Table with name found

https://github.com/nfdi4plants/ARCtrl/blob/aa564a937bd92a13327ad2d80b205012613bb578/src/ARCtrl/Templates/Template.Spreadsheet.fs#L288

### PR DESCRIPTION
**Pull request by Freymaurer**
Starting to write developer docs. Until Fable allows keeping comments, auto generated source code docs do not seem like a valid option.

@JonasLukasczyk please have a look at the js code implementation `/docs/scripts_js/ARC.js` and please just sent me all of your ARCtrl code implementations.

@HLWeil tell me if you are okay with the structur, file placement, etc.

closes #223 
closes #231 
closes #233 
closes #234 
closes #232 
closes #114

### CODE REVIEW

File: docs/ARC.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+- 📁 ARC root
+  - 📄 isa.investigation.xlsx
+  - 📁 workflows
+  - 📁 runs
+  - 📁 assays
+  - 📁 studies
+
+## Write
+
+In .NET you can use [ARCtrl.NET][1] to handle any contract based read/write operations. For this documentation we will extract the relevant ARCtrl.NET functions.
```
Comment (HLWeil) [inline docs/ARC.md:44]: Maybe we actually also add documentation for ARCtrl.Net here?

**Review by HLWeil — CHANGES_REQUESTED** [review]
I think we should add direct links from the main `README.md` to the `.md` files located in the docs folder.
Similar to https://github.com/fslaborg/FSharp.Stats

**Comment by HLWeil** [comment]
Maybe split ARC.md into `IO.md` and `API.md`?

**Comment by Freymaurer** [comment]
@HLWeil can you review this, so we can just release a new beta version and i will add some more docs soon?

**Review by HLWeil — APPROVED** [review]
Good starting point for the documentation. We should keep in mind to reorganize a bit (especially root README.md) when we get a little bit more bulk.

**Comment by Freymaurer** [comment]
> Good starting point for the documentation. We should keep in mind to reorganize a bit (especially root README.md) when we get a little bit more bulk.

Damn! I forgot to push my last changes to README.md!

**Pull request merged**